packages/block-library/src/navigation/edit/use-convert-classic-menu-to-block-menu.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * WordPress dependencies
+ */
+import { useRegistry } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+import {
+	useReducer,
+	useCallback,
+	useRef,
+	useLayoutEffect,
+} from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useCreateNavigationMenu from './use-create-navigation-menu';
+import menuItemsToBlocks from '../menu-items-to-blocks';
+
+function reducer( state, action ) {
+	switch ( action.type ) {
+		case 'RESOLVED':
+			return {
+				...state,
+				isFetching: false,
+				navMenu: action.navMenu,
+			};
+		case 'ERROR':
+			return {
+				...state,
+				isFetching: false,
+				navMenu: null,
+			};
+		case 'LOADING':
+			return {
+				...state,
+				isFetching: true,
+			};
+		default:
+			throw new Error( `Unexpected action type ${ action.type }` );
+	}
+}
+
+function useSafeDispatch( dispatch ) {
+	const mounted = useRef( false );
+	useLayoutEffect( () => {
+		mounted.current = true;
+		return () => ( mounted.current = false );
+	}, [] );
+	return useCallback(
+		( ...args ) => ( mounted.current ? dispatch( ...args ) : void 0 ),
+		[ dispatch ]
+	);
+}
+
+function useConvertClassicToBlockMenu( clientId ) {
+	const createNavigationMenu = useCreateNavigationMenu( clientId );
+	const registry = useRegistry();
+
+	const [ state, dispatch ] = useReducer( reducer, {
+		navMenu: null,
+		isFetching: false,
+	} );
```

Comment:
That seems a bit overcomplicated to me for two state variables and three possible transitions, could this look like this instead?

```js
const [ navMenu, setNavMenu ] = useState();
const [ isFetching, setIsFetching ] = useState();
const setLoading = useCallback( () => {
	setIsFetching( true );
	setNavMenu( null );
} );
const setError = useCallback( () => {
	setIsFetching( false );
	setNavMenu( null );
} );
const setResolved = useCallback( ( menus ) => {
	setIsFetching( false );
	setNavMenu( menus );
} );
```

Reply:
Just noting I proposed this but it was rejected https://github.com/WordPress/gutenberg/pull/38956

Reply:
That error seems to be harmless though?

Reply:
I've had a change of heart. Let's avoid adding complexity where it's not warranted. I've refactored to `useState`.